### ISSUE TITLE
[webhook] Custom responses

### ISSUE THREAD

**Issue by ParkMyCar** [MaterializeInc/materialize#21299]
### Feature request

Certain webhook providers, e.g. [Kinesis Data Firehose](https://aws.amazon.com/kinesis/data-firehose/), expect a certain [response format](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#responseformat). A 200 response is not enough, failure to provide this format prevents the provider from knowing it's request succeeded and various bad behavior ensues, e.g. repeatedly sending the same event.

@sthm tried out DynamoDB -> Kinesis -> Materialize and couldn't get it work because we were not returning the expected response. [Slack](https://materializeinc.slack.com/archives/CNVRXGFDJ/p1692352699618359)

Implementing this shouldn't be too difficult, the hardest part is determining the SQL syntax to customize the response. Like webhook validation, we need to be fairly generic. As a guiding example Firehose expects the following in the body of the response:
```
{
  requestId: Must match the requestId from the headers in the request.
  timestamp: The timestamp (milliseconds since epoch) at which the server processed this request.
}
```

Initial strawman from @benesch:
```
CREATE SOURCE … FROM WEBHOOK
...
SUCCESS RESPONSE HEADER ‘content-encoding’ => ‘application/json’
SUCCESS RESPONSE BODY (
    [WITH HEADER, BODY]
    jsonb_build_object(‘responseId’, body->>’responseId’)
)
```

**Comment by ParkMyCar** [comment]
I want to propose slight modifications to Nikhil's initial strawman that I think aligns this well with the existing syntax:
```
-- A comma separated list of <header name> AS <header value>, similar to the future header 
-- filtering syntax that is proposed in the original design doc.
SUCCESS RESPONSE HEADERS (
  'content-encoding' AS 'appliation/json'
)

-- Require specifying the response body format to align with the request side.
SUCCESS RESPONSE BODY FORMAT [BYTES | TEXT | JSON] (
  [
    -- This WITH clause should align with that from the CHECK statement
    WITH (
      -- Optionally we can provide the headers with an alias
      [HEADERS] [alias],
      -- Optionally we can provide the body parsed as BYTES or JSON (default TEXT) with an alias
      [BODY [AS [BYTES | JSON] [alias]]
    )
  ]
  <expression that returns type specified as BODY FORMAT>
)
```

**Comment by benesch** [comment]
Thanks for filing and iterating on this, @ParkMyCar! A few thoughts.

---

First, the `SUCCESS RESPONSE HEADERS` bit. Here's the strawman proposal from the design doc on how we'd extend `INCLUDE HEADERS` to support including headers one by one, for context:

```
CREATE SOURCE src FROM WEBHOOK (
    INCLUDE HEADERS (
        'header-1' AS header_1
        'header-2' AS header_2 BYTES
    ),
    CHECK (
        WITH (HEADER 'x-signature' AS x_signature)
        x_signature ...
    )
)
```

Note here that what follows `AS` is an identifier—a name for the header. This is consistent with how `AS `works in `SELECT 1 AS col_name`. Note that CTEs (`WITH a AS (SELECT 1), b AS (SELECT 2)`) by reversing the names and definitions... but fundamentally still about assigning a name to a thing.

Using `AS` in `SUCCESS RESPONSE HEADERS` is a bit at odds with existing uses of `AS`, as neither side is an identifier name!

```
SUCCESS RESPONSE HEADERS (
  'content-encoding' AS 'appliation/json'
)
```

My proposal:

```
SUCCESS RESPONSE HEADER ‘content-encoding’ => ‘application/json’
```

is bad for other reasons—the ` => ` syntax is something we made up for our `map` type, which is not widely used—but does avoid overloading `AS`.

tl;dr I think worth iterating on the design of this option a bit more.

----

Re `SUCCESS RESPONSE BODY`, totally agree that the inside of that option should support `WITH` in _exactly_ the same way as the `CHECK` expression.

I don't think we need the `FORMAT`, though! It's a great point that we need to handle different output formats... but I think we can just infer the right output format! SQL expressions always have a knowable return type if all input types are known. So we could just say that `SUCCESS RESPONSE BODY` can have type `bytea` _or_ type `text` _or_ type `jsonb`, and we'll automatically handle the encoding of those types. In fact, we could even set the `content-encoding` header automatically based on the response body type, unless you've explicitly overridden the `content-encoding` header with a `SUCCESS RESPONSE HEADER` option.

**Comment by ParkMyCar** [comment]
Thanks for iterating here @benesch!

For defining headers, what do you think about the `IS` keyword? It's used in Postgres for defining a [`COMMENT`](https://www.postgresql.org/docs/current/sql-comment.html), I think it might align well here! What do you think about supporting `WITH (HEADERS, BODY)` here too? e.g.

```
SUCCESS RESPONSE HEADERS (
  WITH (HEADERS),
  'content-encoding' IS 'application/json',
  'x-request-id' IS headers->'x-request-id'
)
```

Re `SUCCESS RESPONSE BODY`, I think that all makes sense! Also automatically setting `content-encoding` is 👌. So we'd support exactly what you proposed in your initial design,
```
SUCCESS RESPONSE BODY (
    [WITH HEADER, BODY]
    jsonb_build_object(‘responseId’, body->>’responseId’)
)
```
where the final expression must have a return type of `bytea`, `text`, or `jsonb`

**Comment by benesch** [comment]
> What do you think about supporting `WITH (HEADERS, BODY)` here too?

Makes perfect sense!

> For defining headers, what do you think about the `IS` keyword? It's used in Postgres for defining a [`COMMENT`](https://www.postgresql.org/docs/current/sql-comment.html), I think it might align well here!

Eh, seems even weirder than `=>` to me. Points for being used in `COMMENT`, I suppose, but man that is weird syntax. Are we overcomplicating this? Should we just use `=`?

```
SUCCESS RESPONSE HEADERS (
  WITH (HEADERS),
  'content-encoding' = 'application/json',
  'x-request-id' = headers->'x-request-id'
)
```

**Comment by ParkMyCar** [comment]
> Are we overcomplicating this?

We might be 😅

> Should we just use `=`?
>
> ```
> SUCCESS RESPONSE HEADERS (
>  WITH (HEADERS),
>  'content-encoding' = 'application/json',
>  'x-request-id' = headers->'x-request-id'
>)
> ```

That works for me!

**Comment by joacoc** [comment]
+1 for `=`. Felt it as the most natural approach.
